File: patterns/StrategyPattern9.py
#!/usr/bin/env python3

#pylint: disable = bad-builtin, no-self-use, too-few-public-methods, unused-argument

# ---------
# Store9.py
# ---------

class Price :
    def get_points (self, days_rented) : # const
        return 1

class RegularPrice (Price) :
    def get_amount (self, days_rented) : # const
        result = 2
        if days_rented > 2 :
            result += (days_rented - 2) * 1.5
        return result

class NewReleasePrice (Price) :
    def get_amount (self, days_rented) : # const
        return days_rented * 3

    def get_points (self, days_rented) : # const
        return 2 if (days_rented > 1) else 1

class ChildrensPrice (Price) :
    def get_amount (self, days_rented) : # const
        result = 1.5
        if days_rented > 3 :
            result += (days_rented - 3) * 1.5
        return result

class Movie :
    REGULAR     = "RegularPrice"
    NEW_RELEASE = "NewReleasePrice"
    CHILDRENS   = "ChildrensPrice"

    def __init__ (self, title, price) :
        self.title = title
        self.set_price(price)

    """
    price
        get_amount()
    """
    def get_amount (self, days_rented) : # const
        return self.price.get_amount(days_rented)

    """
    price
        get_points()
    """
    def get_points (self, days_rented) :
        return self.price.get_points(days_rented)

    def get_output (self, days_rented) : # const
        return "\t" + self.title + "\t" + str(self.get_amount(days_rented)) + "\n"

    def get_title (self) : # const
        return self.title

    def set_price (self, price) :
        try :
            self.price = globals()[price]()
        except KeyError as e:
            print(e.args)
        except TypeError as e:
            print(e.args)

class Rental :
    def __init__ (self, movie, days_rented) :
        self.movie       = movie
        self.days_rented = days_rented

    """
    movie()
        get_amount()
    """
    def get_amount (self) : # const
        return self.movie.get_amount(self.days_rented)

    def get_days_rented (self) : # const # no longer used
        return self.days_rented

    def get_movie (self) : # const
        return self.movie

    """
    movie
        get_output()
    """
    def get_output (self) : # const
        return self.movie.get_output(self.days_rented)

    """
    movie
        get_points()
    """
    def get_points (self) : # const
        return self.movie.get_points(self.days_rented)

class Customer :
    def __init__ (self, name) :
        self.name    = name
        self.rentals = []

    def add_rental (self, rental) :
        self.rentals.append(rental)

    def get_name (self) : # const
        return self.name

    """
    iter(rentals).next()
        get_amount()
        get_points()
        get_output()
    """
    def statement (self) : # O(n)
        amount  = sum(map(Rental.get_amount, self.rentals))
        points  = sum(map(Rental.get_points, self.rentals))
        result  =                  \
            "Rental Record for " + \
            self.get_name()      + \
            "\n"                 + \
            "".join(map(Rental.get_output, self.rentals))
        result += "Amount owed is " + str(amount) + "\n"
        result += "You earned "     + str(points) + " frequent renter points"
        return result

print("Store9.py")

x = Customer("Penelope")
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "Amount owed is 0\n"                  + \
    "You earned 0 frequent renter points"

x.add_rental(Rental(Movie("Shane", Movie.REGULAR), 2))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "Amount owed is 2\n"                  + \
    "You earned 1 frequent renter points"

x.add_rental(Rental(Movie("Red River", Movie.REGULAR), 5))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "\tRed River\t6.5\n"                  + \
    "Amount owed is 8.5\n"                + \
    "You earned 2 frequent renter points"

x.add_rental(Rental(Movie("Giant", Movie.NEW_RELEASE), 1))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "\tRed River\t6.5\n"                  + \
    "\tGiant\t3\n"                        + \
    "Amount owed is 11.5\n"               + \
    "You earned 3 frequent renter points"

x.add_rental(Rental(Movie("2001", Movie.NEW_RELEASE), 3))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "\tRed River\t6.5\n"                  + \
    "\tGiant\t3\n"                        + \
    "\t2001\t9\n"                         + \
    "Amount owed is 20.5\n"               + \
    "You earned 5 frequent renter points"

x.add_rental(Rental(Movie("Big Country", Movie.CHILDRENS), 3))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "\tRed River\t6.5\n"                  + \
    "\tGiant\t3\n"                        + \
    "\t2001\t9\n"                         + \
    "\tBig Country\t1.5\n"                + \
    "Amount owed is 22.0\n"               + \
    "You earned 6 frequent renter points"

x.add_rental(Rental(Movie("Spartacus", Movie.CHILDRENS), 5))
assert x.statement() ==                    \
    "Rental Record for Penelope\n"        + \
    "\tShane\t2\n"                        + \
    "\tRed River\t6.5\n"                  + \
    "\tGiant\t3\n"                        + \
    "\t2001\t9\n"                         + \
    "\tBig Country\t1.5\n"                + \
    "\tSpartacus\t4.5\n"                  + \
    "Amount owed is 26.5\n"               + \
    "You earned 7 frequent renter points"

print("Done.")

File: patterns/StrategyPattern9.java
// -----------
// Store9.java
// -----------

/*
Change Price to an interface
Create AbstractPrice
Create Movie.getOutput()
Create Rental.getOutput()
Remove Customer.getTotalCharge()
Remove Customer.getTotalFrequentRenterPoints()
Rename getCharge()               to getAmount().
Rename getFrequentRenterPoints() to getPoints()
*/

import java.util.Enumeration;
import java.util.Vector;

interface Price {
    double getAmount    (int daysRented);
    int    getPoints    (int daysRented);
    String getPriceCode ();}              // not used

abstract class AbstractPrice implements Price {
    public int getPoints (int daysRented) { // const
        return 1;}}

class RegularPrice extends AbstractPrice {
    public double getAmount (int daysRented) { // const
        double result = 2;
        if (daysRented > 2)
            result += (daysRented - 2) * 1.5;
        return result;}

    public String getPriceCode () { // const, not used
        return Movie.REGULAR;}}

class NewReleasePrice extends AbstractPrice {
    public double getAmount (int daysRented) { // const
        return daysRented * 3;}

    public int getPoints (int daysRented) { // const
        return (daysRented > 1) ? 2 : 1;}

    public String getPriceCode () { // const, not used
        return Movie.NEW_RELEASE;}}

class ChildrensPrice extends AbstractPrice {
    public double getAmount (int daysRented) { // const
        double result = 1.5;
        if (daysRented > 3)
            result += (daysRented - 3) * 1.5;
        return result;}

    public String getPriceCode () { // const, not used
        return Movie.CHILDRENS;}}

class Movie {
    public static final String REGULAR     = "RegularPrice";
    public static final String NEW_RELEASE = "NewReleasePrice";
    public static final String CHILDRENS   = "ChildrensPrice";

    private String _title;
    private Price  _price;

    public Movie (String title, String priceCode) {
        _title = title;
        setPriceCode(priceCode);}

    /**
     * _price
     *     getAmount()
     */
    public double getAmount (int daysRented) { // const
        return _price.getAmount(daysRented);}

    public String getOutput (int daysRented) { // const
        return
            "\t" + _title                                +
            "\t" + String.valueOf(getAmount(daysRented)) + "\n";}

    /**
     * _price
     *     getPoints()
     */
    public int getPoints (int daysRented) { // const
        return _price.getPoints(daysRented);}

    /**
     * _price
     *     getPriceCode()
     */
    public String getPriceCode () { // const
        return _price.getPriceCode();}

    public String getTitle () { // const
        return _title;}

    public void setPriceCode (String priceCode) {
        try {
            _price = (Price) Class.forName(priceCode).newInstance();}
        catch (ClassCastException e) {
            throw new IllegalArgumentException("Incorrect Price Code");}
        catch (ClassNotFoundException e) {
            throw new IllegalArgumentException("Incorrect Price Code");}
        catch (IllegalAccessException e) {
            throw new IllegalArgumentException("Incorrect Price Code");}
        catch (InstantiationException e) {
            throw new IllegalArgumentException("Incorrect Price Code");}}}

class Rental {
    private Movie _movie;
    private int   _daysRented;

    public Rental (Movie movie, int daysRented) {
        _movie      = movie;
        _daysRented = daysRented;}

    /**
     * _movie
     *     getAmount()
     */
    public double getAmount () { // const
        return _movie.getAmount(_daysRented);}

    public int getDaysRented () { // const // no longer used
        return _daysRented;}

    public Movie getMovie () { // const
        return _movie;}

    /**
     * _movie
     *     getOutput()
     */
    public String getOutput () { // const
        return _movie.getOutput(_daysRented);}

    /**
     * _movie
     *     getPoints()
     */
    public int getPoints () { // const
        return _movie.getPoints(_daysRented);}}

class Customer {
    private String         _name;
    private Vector<Rental> _rentals = new Vector<Rental>();

    public Customer (String name) {
        _name = name;}

    public void addRental (Rental rental) {
        _rentals.addElement(rental);}

    public String getName () { // const
        return _name;}

    /**
     * _rentals.elements().nextElement()
     *     getAmount()
     *     getMovie()
     *         getTitle()
     */
    public String statement () { // O(3n)
        double amount = 0;
        for (Rental rental : _rentals)
            amount += rental.getAmount();
        int points = 0;
        for (Rental rental : _rentals)
            points += rental.getPoints();
        String result = "Rental Record for " + getName() + "\n";
        for (Rental rental : _rentals)
            result += rental.getOutput();
        result += "Amount owed is " + String.valueOf(amount) + "\n";
        result += "You earned " + String.valueOf(points) + " frequent renter points";
        return result;}}

final class Store9 {
    public static void main (String[] args) {
        System.out.println("Store9.java");

        Customer x = new Customer("Penelope");
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "Amount owed is 0.0\n"         +
            "You earned 0 frequent renter points");

        x.addRental(new Rental(new Movie("Shane", Movie.REGULAR), 2));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "Amount owed is 2.0\n"         +
            "You earned 1 frequent renter points");

        x.addRental(new Rental(new Movie("Red River", Movie.REGULAR), 5));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "\tRed River\t6.5\n"           +
            "Amount owed is 8.5\n"         +
            "You earned 2 frequent renter points");

        x.addRental(new Rental(new Movie("Giant", Movie.NEW_RELEASE), 1));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "\tRed River\t6.5\n"           +
            "\tGiant\t3.0\n"               +
            "Amount owed is 11.5\n"        +
            "You earned 3 frequent renter points");

        x.addRental(new Rental(new Movie("2001", Movie.NEW_RELEASE), 3));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "\tRed River\t6.5\n"           +
            "\tGiant\t3.0\n"               +
            "\t2001\t9.0\n"                +
            "Amount owed is 20.5\n"        +
            "You earned 5 frequent renter points");

        x.addRental(new Rental(new Movie("Big Country", Movie.CHILDRENS), 3));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "\tRed River\t6.5\n"           +
            "\tGiant\t3.0\n"               +
            "\t2001\t9.0\n"                +
            "\tBig Country\t1.5\n"         +
            "Amount owed is 22.0\n"        +
            "You earned 6 frequent renter points");

        x.addRental(new Rental(new Movie("Spartacus", Movie.CHILDRENS), 5));
        assert x.statement().equals(
            "Rental Record for Penelope\n" +
            "\tShane\t2.0\n"               +
            "\tRed River\t6.5\n"           +
            "\tGiant\t3.0\n"               +
            "\t2001\t9.0\n"                +
            "\tBig Country\t1.5\n"         +
            "\tSpartacus\t4.5\n"           +
            "Amount owed is 26.5\n"        +
            "You earned 7 frequent renter points");

        System.out.println("Done.");}}
